# Netflix Dataset Preparation - Milestone 1
## Week 1 & 2: Requirements & Dataset Preparation

**Project Goals:**
- Define project scope and success metrics
- Load the Netflix Kaggle dataset
- Clean the dataset (handle missing values, remove duplicates)
- Normalize categorical features such as genre, rating, and country


## 1. Import Required Libraries
First, we'll import the necessary libraries for data manipulation and analysis.

In [ ]:
import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')

print("✓ Libraries imported successfully!")

: 

## 2. Load the Netflix Dataset
Load the Netflix titles dataset from the CSV file.

In [ ]:
def load_dataset(filepath='C:\Users\Kavya Jain\OneDrive\Desktop\Project1\netflix\Data\netflix_titles.csv'):
    """Load the Netflix dataset from CSV file."""
    print("=" * 70)
    print("LOADING DATASET")
    print("=" * 70)
    
    try:
        df = pd.read_csv(filepath)
        print(f"✓ Dataset loaded successfully!")
        print(f"  - Shape: {df.shape[0]} rows × {df.shape[1]} columns")
        print(f"  - Columns: {list(df.columns)}")
        return df
    except FileNotFoundError:
        print(f"✗ Error: File '{filepath}' not found!")
        return None
    except Exception as e:
        print(f"✗ Error loading dataset: {str(e)}")
        return None

# Load the dataset
df = load_dataset('netflix_titles.csv')

: 

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   show_id       8807 non-null   str  
 1   type          8807 non-null   str  
 2   title         8807 non-null   str  
 3   director      6173 non-null   str  
 4   cast          7982 non-null   str  
 5   country       7976 non-null   str  
 6   date_added    8797 non-null   str  
 7   release_year  8807 non-null   int64
 8   rating        8803 non-null   str  
 9   duration      8804 non-null   str  
 10  listed_in     8807 non-null   str  
 11  description   8807 non-null   str  
dtypes: int64(1), str(11)
memory usage: 825.8 KB


## 3. Data Quality Assessment (Before Cleaning)
Assess the data quality to identify missing values, duplicates, and data types.

In [3]:
def assess_data_quality(df):
    """Assess data quality before cleaning."""
    print("\n" + "=" * 70)
    print("DATA QUALITY ASSESSMENT")
    print("=" * 70)
    
    print("\n1. Dataset Overview:")
    print(f"   Total Records: {len(df)}")
    print(f"   Total Columns: {len(df.columns)}")
    
    print("\n2. Missing Values:")
    missing = df.isnull().sum()
    missing_pct = (missing / len(df) * 100).round(2)
    missing_df = pd.DataFrame({
        'Missing Count': missing,
        'Percentage': missing_pct
    })
    print(missing_df[missing_df['Missing Count'] > 0])
    
    print("\n3. Duplicate Records:")
    duplicates = df.duplicated().sum()
    print(f"   Total Duplicates: {duplicates}")
    
    print("\n4. Data Types:")
    print(df.dtypes)
    
    return {
        'total_records': len(df),
        'missing_values': missing.sum(),
        'duplicates': duplicates
    }

# Assess data quality
original_quality = assess_data_quality(df)


DATA QUALITY ASSESSMENT

1. Dataset Overview:
   Total Records: 8807
   Total Columns: 12

2. Missing Values:
            Missing Count  Percentage
director             2634       29.91
cast                  825        9.37
country               831        9.44
date_added             10        0.11
rating                  4        0.05
duration                3        0.03

3. Duplicate Records:
   Total Duplicates: 0

4. Data Types:
show_id           str
type              str
title             str
director          str
cast              str
country           str
date_added        str
release_year    int64
rating            str
duration          str
listed_in         str
description       str
dtype: object


## 4. Clean the Dataset
Handle missing values and remove duplicate records.

In [5]:
def clean_dataset(df):
    """Clean the dataset: handle missing values and remove duplicates."""
    print("\n" + "=" * 70)
    print("CLEANING DATASET")
    print("=" * 70)
    
    df_clean = df.copy()
    
    # Remove duplicate records
    print("\n1. Removing Duplicates...")
    initial_count = len(df_clean)
    df_clean = df_clean.drop_duplicates()
    duplicates_removed = initial_count - len(df_clean)
    print(f"   ✓ Removed {duplicates_removed} duplicate records")
    
    # Handle missing values based on data quality assessment
    print("\n2. Handling Missing Values...")
    
    # High missing values (>29%): director
    if 'director' in df_clean.columns:
        missing_before = df_clean['director'].isnull().sum()
        if missing_before > 0:
            df_clean['director'] = df_clean['director'].fillna('Unknown')
            print(f"   ✓ director: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Unknown'")
    
    # Moderate missing values (5-10%): cast, country
    if 'cast' in df_clean.columns:
        missing_before = df_clean['cast'].isnull().sum()
        if missing_before > 0:
            df_clean['cast'] = df_clean['cast'].fillna('Unknown')
            print(f"   ✓ cast: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Unknown'")
    
    if 'country' in df_clean.columns:
        missing_before = df_clean['country'].isnull().sum()
        if missing_before > 0:
            df_clean['country'] = df_clean['country'].fillna('Unknown')
            print(f"   ✓ country: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Unknown'")
    
    # Low missing values (<1%): date_added, rating, duration
    if 'date_added' in df_clean.columns:
        missing_before = df_clean['date_added'].isnull().sum()
        if missing_before > 0:
            df_clean['date_added'] = df_clean['date_added'].fillna('Unknown')
            print(f"   ✓ date_added: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Unknown'")
    
    if 'rating' in df_clean.columns:
        missing_before = df_clean['rating'].isnull().sum()
        if missing_before > 0:
            df_clean['rating'] = df_clean['rating'].fillna('Not Rated')
            print(f"   ✓ rating: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Not Rated'")
    
    if 'duration' in df_clean.columns:
        missing_before = df_clean['duration'].isnull().sum()
        if missing_before > 0:
            df_clean['duration'] = df_clean['duration'].fillna('Unknown')
            print(f"   ✓ duration: Filled {missing_before} missing values ({(missing_before/len(df_clean)*100):.2f}%) with 'Unknown'")
    
    # Handle other potential missing values
    if 'listed_in' in df_clean.columns:
        missing_before = df_clean['listed_in'].isnull().sum()
        if missing_before > 0:
            df_clean['listed_in'] = df_clean['listed_in'].fillna('Uncategorized')
            print(f"   ✓ listed_in: Filled {missing_before} missing values with 'Uncategorized'")
    
    if 'description' in df_clean.columns:
        missing_before = df_clean['description'].isnull().sum()
        if missing_before > 0:
            df_clean['description'] = df_clean['description'].fillna('No description available')
            print(f"   ✓ description: Filled {missing_before} missing values with 'No description available'")
    
    # Check for any remaining missing values
    remaining_missing = df_clean.isnull().sum().sum()
    if remaining_missing == 0:
        print(f"\n   ✓ All missing values handled successfully!")
    else:
        print(f"\n   ⚠ Warning: {remaining_missing} missing values remain in other columns")
    
    print(f"   ✓ Final Record Count: {len(df_clean)}")
    
    return df_clean

# Clean the dataset
df_cleaned = clean_dataset(df)


CLEANING DATASET

1. Removing Duplicates...
   ✓ Removed 0 duplicate records

2. Handling Missing Values...
   ✓ director: Filled 2634 missing values (29.91%) with 'Unknown'
   ✓ cast: Filled 825 missing values (9.37%) with 'Unknown'
   ✓ country: Filled 831 missing values (9.44%) with 'Unknown'
   ✓ date_added: Filled 10 missing values (0.11%) with 'Unknown'
   ✓ rating: Filled 4 missing values (0.05%) with 'Not Rated'
   ✓ duration: Filled 3 missing values (0.03%) with 'Unknown'

   ✓ All missing values handled successfully!
   ✓ Final Record Count: 8807


## 5. Normalize Categorical Features
Standardize genre, rating, and country columns.

In [6]:
def normalize_categorical_features(df):
    """Normalize genre, rating, and country columns."""
    print("\n" + "=" * 70)
    print("NORMALIZING CATEGORICAL FEATURES")
    print("=" * 70)
    
    df_normalized = df.copy()
    
    # Normalize Genre (listed_in)
    if 'listed_in' in df_normalized.columns:
        print("\n1. Normalizing Genre (listed_in)...")
        df_normalized['listed_in'] = df_normalized['listed_in'].str.strip()
        df_normalized['genre_list'] = df_normalized['listed_in'].apply(
            lambda x: [g.strip() for g in str(x).split(',')] if pd.notnull(x) else []
        )
        unique_genres = df_normalized['listed_in'].nunique()
        print(f"   ✓ Genres normalized and standardized")
        print(f"   ✓ Unique genre combinations: {unique_genres}")
        
        all_genres = []
        for genres in df_normalized['genre_list']:
            all_genres.extend(genres)
        unique_individual = len(set(all_genres))
        print(f"   ✓ Unique individual genres: {unique_individual}")
    
    # Normalize Rating
    if 'rating' in df_normalized.columns:
        print("\n2. Normalizing Rating...")
        df_normalized['rating'] = df_normalized['rating'].str.strip().str.upper()
        unique_ratings = df_normalized['rating'].nunique()
        print(f"   ✓ Ratings normalized to uppercase")
        print(f"   ✓ Unique ratings: {unique_ratings}")
        print(f"   ✓ Rating categories: {sorted(df_normalized['rating'].unique())}")
    
    # Normalize Country
    if 'country' in df_normalized.columns:
        print("\n3. Normalizing Country...")
        df_normalized['country'] = df_normalized['country'].str.strip()
        df_normalized['country_list'] = df_normalized['country'].apply(
            lambda x: [c.strip() for c in str(x).split(',')] if pd.notnull(x) else []
        )
        df_normalized['primary_country'] = df_normalized['country_list'].apply(
            lambda x: x[0] if len(x) > 0 else 'Unknown'
        )
        unique_countries = df_normalized['primary_country'].nunique()
        print(f"   ✓ Countries normalized and standardized")
        print(f"   ✓ Unique primary countries: {unique_countries}")
        
        all_countries = []
        for countries in df_normalized['country_list']:
            all_countries.extend(countries)
        unique_individual = len(set(all_countries))
        print(f"   ✓ Total unique countries (including co-productions): {unique_individual}")
    
    # Normalize Content Type
    if 'type' in df_normalized.columns:
        print("\n4. Normalizing Content Type...")
        df_normalized['type'] = df_normalized['type'].str.strip()
        content_types = df_normalized['type'].value_counts()
        print(f"   ✓ Content types: {dict(content_types)}")
    
    return df_normalized

# Normalize categorical features
df_normalized = normalize_categorical_features(df_cleaned)


NORMALIZING CATEGORICAL FEATURES

1. Normalizing Genre (listed_in)...
   ✓ Genres normalized and standardized
   ✓ Unique genre combinations: 514
   ✓ Unique individual genres: 42

2. Normalizing Rating...
   ✓ Ratings normalized to uppercase
   ✓ Unique ratings: 18
   ✓ Rating categories: ['66 MIN', '74 MIN', '84 MIN', 'G', 'NC-17', 'NOT RATED', 'NR', 'PG', 'PG-13', 'R', 'TV-14', 'TV-G', 'TV-MA', 'TV-PG', 'TV-Y', 'TV-Y7', 'TV-Y7-FV', 'UR']

3. Normalizing Country...
   ✓ Countries normalized and standardized
   ✓ Unique primary countries: 87
   ✓ Total unique countries (including co-productions): 124

4. Normalizing Content Type...
   ✓ Content types: {'Movie': np.int64(6131), 'TV Show': np.int64(2676)}


## 6. Generate Data Quality Report
Generate a comprehensive report showing the improvements after cleaning.

In [7]:
def generate_quality_report(df_original, df_cleaned):
    """Generate final data quality report."""
    print("\n" + "=" * 70)
    print("DATA QUALITY REPORT (AFTER CLEANING)")
    print("=" * 70)
    
    print("\n1. Cleaning Summary:")
    print(f"   Original Records: {len(df_original)}")
    print(f"   Final Records: {len(df_cleaned)}")
    print(f"   Records Removed: {len(df_original) - len(df_cleaned)}")
    
    print("\n2. Missing Values After Cleaning:")
    missing = df_cleaned.isnull().sum()
    if missing.sum() == 0:
        print("   ✓ No missing values remaining!")
    else:
        print(missing[missing > 0])
    
    print("\n3. Data Completeness:")
    completeness = ((len(df_cleaned) - df_cleaned.isnull().sum()) / len(df_cleaned) * 100).round(2)
    for col in df_cleaned.columns:
        print(f"   {col}: {completeness[col]}%")
    
    print("\n4. Key Statistics:")
    if 'type' in df_cleaned.columns:
        print(f"   Content Types: {df_cleaned['type'].value_counts().to_dict()}")
    if 'rating' in df_cleaned.columns:
        print(f"   Top 5 Ratings: {df_cleaned['rating'].value_counts().head().to_dict()}")
    if 'primary_country' in df_cleaned.columns:
        print(f"   Top 5 Countries: {df_cleaned['primary_country'].value_counts().head().to_dict()}")
    
    return {
        'original_records': len(df_original),
        'final_records': len(df_cleaned),
        'completeness': completeness.mean()
    }

# Generate quality report
final_quality = generate_quality_report(df, df_normalized)


DATA QUALITY REPORT (AFTER CLEANING)

1. Cleaning Summary:
   Original Records: 8807
   Final Records: 8807
   Records Removed: 0

2. Missing Values After Cleaning:
   ✓ No missing values remaining!

3. Data Completeness:
   show_id: 100.0%
   type: 100.0%
   title: 100.0%
   director: 100.0%
   cast: 100.0%
   country: 100.0%
   date_added: 100.0%
   release_year: 100.0%
   rating: 100.0%
   duration: 100.0%
   listed_in: 100.0%
   description: 100.0%
   genre_list: 100.0%
   country_list: 100.0%
   primary_country: 100.0%

4. Key Statistics:
   Content Types: {'Movie': 6131, 'TV Show': 2676}
   Top 5 Ratings: {'TV-MA': 3207, 'TV-14': 2160, 'TV-PG': 863, 'R': 799, 'PG-13': 490}
   Top 5 Countries: {'United States': 3211, 'India': 1008, 'Unknown': 831, 'United Kingdom': 628, 'Canada': 271}


## 7. Save Cleaned Dataset
Save the cleaned and normalized dataset to a new CSV file.

In [8]:
def save_cleaned_dataset(df, output_filename='netflix_titles_cleaned.csv'):
    """Save the cleaned and normalized dataset."""
    print("\n" + "=" * 70)
    print("SAVING CLEANED DATASET")
    print("=" * 70)
    
    try:
        df.to_csv(output_filename, index=False)
        print(f"✓ Cleaned dataset saved as '{output_filename}'")
        print(f"  - Shape: {df.shape[0]} rows × {df.shape[1]} columns")
        return True
    except Exception as e:
        print(f"✗ Error saving dataset: {str(e)}")
        return False

# Save the cleaned dataset
save_cleaned_dataset(df_normalized)


SAVING CLEANED DATASET
✓ Cleaned dataset saved as 'netflix_titles_cleaned.csv'
  - Shape: 8807 rows × 15 columns


True

## 8. Summary and Next Steps

### ✅ Milestone 1 Completed!

**What we accomplished:**
- ✓ Loaded 8,807 Netflix titles (6,131 movies + 2,676 TV shows)
- ✓ Cleaned dataset: Handled 4,307 missing values across 6 columns
- ✓ Achieved 100% data completeness
- ✓ Normalized categorical features:
  - **Genre:** 42 unique individual genres
  - **Rating:** 18 rating categories
  - **Country:** 124 countries
- ✓ Created enhanced dataset with 3 additional columns (genre_list, country_list, primary_country)

**Output Files:**
- `netflix_titles_cleaned.csv` - Clean dataset ready for analysis

**Next Steps:**
- Exploratory Data Analysis (EDA)
- Visualization of trends and patterns
- Statistical analysis
- Potential modeling (recommendation systems, predictions, etc.)

In [9]:
# Preview the cleaned dataset
print("Preview of Cleaned Dataset:")
print("\nFirst 5 rows:")
print(df_normalized.head())
print("\nDataset Info:")
print(df_normalized.info())
print("\nBasic Statistics:")
print(df_normalized.describe(include='all'))

Preview of Cleaned Dataset:

First 5 rows:
  show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water          Unknown   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans          Unknown   
4      s5  TV Show           Kota Factory          Unknown   

                                                cast        country  \
0                                            Unknown  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...        Unknown   
3                                            Unknown        Unknown   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA 